foursquare/qgis-plugin
# Support logarithmic scale (UN-5192)

kepler.gl and Studio don't yet support logarithmic scale, so what we end up doing is detecting when a logarithmic graduated symbol is being used in QGIS and then we convert it to `'custom'` (custom breaks) color scale for Studio. We also extract value ranges/classes and provide them in `colorMap`.

<p>
<img src="https://github.com/foursquare/qgis-plugin/assets/1355455/8d004494-39f3-4659-ab2c-d6d83a736893" height="100" />
<img src="https://github.com/foursquare/qgis-plugin/assets/1355455/c921e46c-0de3-4765-a6cf-80e9b1ede345" height="100" />
</p>

**Testing**

- install the plugin from ZIP [file](https://github.com/foursquare/qgis-plugin/files/11505417/Unfolded-dev.zip) (packaged from `9ce44eb` commit) via <kbd>Plugins</kbd> → <kbd>Manage & Install Plugins...</kbd> → <kbd>Install from ZIP</kbd> (notice this installs it under `Unfolded-dev` name)
- open some project in QGIS
- right click on one of the layers → <kbd>Properties...</kbd> → <kbd>Symbology</kbd> section
- pick <kbd>Graduated</kbd> in the top dropdown
- pick "Value" field for which we're making a new scale/graduation
- select "Mode": <kbd>Logarithmic scale</kbd>
- adjust the lower and upper values, colors and number of classes
- close the Properties window
- use the Unfolded plugin to export the map and then import it in Studio
- verify it's showing `Custom Breaks` scale and all the appropriate value ranges and colors are correct and matching as within QGIS
- verify maps in QGIS and Studio match visually

**Notes**

`'custom'` in keplergl uses only upper values and assumes the lower values for its `colorMap`, so we don't specify lower values explicitly, just the upper ones.

Logarithmic graduation in QGIS is not a runtime thing, just a helper on how to specify initial graduation classes (ranges), but the user can still modify them manually after the classification was done. That's why, logarithmic graduation corresponds to custom breaks scale in Studio.

Unresolved questions/notes (regarding the entire codebase, not so much this PR specifically):

- Q: we do `styles = [self._extract_layer_style(category.symbol()) for category in renderer.categories()]`, but we just use `styles[0]`; what about the rest?
  - A: ...
- Q: why only the first one in `symbol.symbolLayers()[0]`?
  - A: it seems we just work with one and don't consider more (because it's also not supported by Studio)
- Q: why this check `if symbol_layer.subSymbol()`?
  - A: symbol is actually a symbol tree (https://docs.qgis.org/3.28/en/docs/user_manual/style_library/symbol_selector.html)
- setting up `mypy` showed so many different errors/warnings, and I've adjusted some bigger offenders to narrow the types appropriately, but some other plainly mutate data types in a way that it is not clear to me what the actual assumption or correct expectation should've been. We'll likely need to clean up this if we intend to maintain this codebase over time.
